mixer/adapter/svcctrl/expiringLRUCache.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package svcctrl
+
+import (
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	"github.com/jonboulle/clockwork"
+)
+
+type (
+	cacheValue struct {
+		// insert timestamp
+		timestamp time.Time
+		value     interface{}
+	}
+
+	// A thread safe LRU cache with expiration.
+	expiringLRUCache struct {
+		mutex      sync.RWMutex
+		lru        *simplelru.LRU
+		clock      clockwork.Clock
+		expiration time.Duration
+	}
+)
+
+func (c *expiringLRUCache) add(key, value interface{}) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.lru.Add(key, &cacheValue{
+		c.clock.Now(),
+		value,
+	})
+}
+
+func (c *expiringLRUCache) get(key interface{}) (interface{}, bool) {
+	value := c.getImpl(key)
+	if value == nil {
+		return nil, false
+	}
+
+	if c.clock.Since(value.timestamp) <= c.expiration {
+		return value.value, true
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
```

Comment:
Can we just call remove(key) and return nil, false?

There could be race condition, but do we really care?

Reply:
simple remove(key) is fine, but my version takes care of race that a new item is inserted between the gap.

Reply:
It is OK to keep it this way.